largescaleobserved.qmd

```diff
@@ -0,0 +1,508 @@
+759200511  2018-09-26 16:31   ml-latest/ratings.csv
+     9784  2018-09-26 16:46   ml-latest/README.txt
+414851573  2018-09-26 16:44   ml-latest/genome-scores.csv
+  2858223  2018-09-26 16:35   ml-latest/movies.csv
+---------                     -------
+1217950223                     8 files
+```
+
+## Structure of the data
+
+One of the purposes of this chapter is to study which dimensions of the data have the greatest effect on the amount of memory used to represent the model and the time required to fit a model.
+
+The two datasets examined in @sec-largescaledesigned, from the English Lexicon Project [@Balota_2007], consist of `trials` or `instances` associated with `subject` and `item` factors.
+The `subject` and `item` factors are "incompletely crossed" in that each item occurred in trials with several subjects, but not all subjects, and each subject responded to several different items, but not to all items.
+
+Similarly in the movie ratings, each instance of a rating is associated with a user and a movie and these factors are incompletely crossed.
+
+```{julia}
+ratings = Arrow.Table(datadir("ratings.arrow"))
+```
+
+Convert this table to a `DataFrame` and drop the `timestamp` column, which we won't be using.
+(For small data sets dropping such columns is not important but with over 27 million ratings it does help to drop unnecessary columns to save some memory space.)
+
+```{julia}
+ratings = select!(DataFrame(ratings), Not(:timestamp));
+```
+
+Information on the movies is available as
+
+```{julia}
+movies = disallowmissing!(
+  leftjoin!(
+    combine(groupby(ratings, :movieId), :rating => mean => :mnrtng),
+    DataFrame(Arrow.Table(datadir("movies.arrow")));
+    on=:movieId,
+  );
+  error=false,
+)
+```
+
+In contrast to data from a designed experiment, like the English Lexicon Project, the data from this observational study are extremely unbalanced with respect to the observational grouping factors, `userId` and `movieId`.
+The `movies` table includes an `nrtngs` column that gives the number of ratings for each movie, which varies from 1 to nearly 100,000.
+
+```{julia}
+extrema(movies.nrtngs)
+```
+
+The number of ratings per user is also highly skewed
+
+```{julia}
+users = combine(
+  groupby(ratings, :userId),
+  :userId => (x -> Int32(length(x))) => :urtngs,
+)
+extrema(users.urtngs)  # minimum and maximum number of ratings per user
+```
+
+One way of visualizing the imbalance in the number of ratings per movie or per user is as an [empirical cumulative distribution function](https://en.wikipedia.org/wiki/Empirical_distribution_function) (ecdf) plot, which is a "stair-step" plot where the vertical axis is the proportion of observations less than or equal to the corresponding value on the horizontal axis.
+Because the distribution of the number of ratings per movie or per user is so highly skewed in the low range we use a logarithmic horizontal axis in @fig-nrtngsecdf.
+
+```{julia}
+#| code-fold: true
```

Comment:
I get a weird error on this block:
```julia
ERROR: ArgumentError: range(1.0, stop=81491.0, length=1): endpoints differ
Stacktrace:
  [1] _linspace1(#unused#::Type{Float64}, start::Float64, stop::Float64, len::Int64)
    @ Base ./twiceprecision.jl:757
  [2] _linspace(#unused#::Type{Float64}, start_n::Int32, stop_n::Int32, len::Int32, den::Int32)
    @ Base ./twiceprecision.jl:737
  [3] _linspace
    @ ./twiceprecision.jl:734 [inlined]
  [4] range_start_stop_length
    @ ./range.jl:571 [inlined]
  [5] _range
    @ ./range.jl:157 [inlined]
  [6] #range#69
    @ ./range.jl:140 [inlined]
  [7] ecdf_xvalues(ecdf::StatsBase.ECDF{Vector{Int32}, StatsBase.Weights{Float64, Float64, Vector{Float64}}}, npoints::Int32)
    @ Makie ~/.julia/packages/Makie/Ppzqh/src/stats/ecdf.jl:5
  [8] convert_arguments(P::Type{Combined{Makie.stairs}}, ecdf::StatsBase.ECDF{Vector{Int32}, StatsBase.Weights{Float64, Float64, Vector{Float64}}}; npoints::Int32)
    @ Makie ~/.julia/packages/Makie/Ppzqh/src/stats/ecdf.jl:11
  [9] #convert_arguments#773
    @ ~/.julia/packages/Makie/Ppzqh/src/stats/ecdf.jl:61 [inlined]
 [10] plot!(scene::Scene, P::Type{Combined{Makie.ecdfplot}}, attributes::Attributes, args::Vector{Int32}; kw_attributes::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Makie ~/.julia/packages/Makie/Ppzqh/src/interfaces.jl:317
 [11] plot!(scene::Scene, P::Type{Combined{Makie.ecdfplot}}, attributes::Attributes, args::Vector{Int32})
    @ Makie ~/.julia/packages/Makie/Ppzqh/src/interfaces.jl:302
 [12] plot(P::Type{Combined{Makie.ecdfplot}}, gp::GridPosition, args::Vector{Int32}; axis::NamedTuple{(:xlabel, :xminorgridvisible, :xminorticks, :xminorticksvisible, :xscale, :ylabel, :yminorticksvisible), Tuple{String, Bool, IntervalsBetween, Bool, typeof(log10), String, Bool}}, kwargs::Base.Pairs{Symbol, Int32, Tuple{Symbol}, NamedTuple{(:npoints,), Tuple{Int32}}})
    @ Makie ~/.julia/packages/Makie/Ppzqh/src/figureplotting.jl:95
 [13] ecdfplot(::GridPosition, ::Vararg{Any}; attributes::Base.Pairs{Symbol, Any, Tuple{Symbol, Symbol}, NamedTuple{(:npoints, :axis), Tuple{Int32, NamedTuple{(:xlabel, :xminorgridvisible, :xminorticks, :xminorticksvisible, :xscale, :ylabel, :yminorticksvisible), Tuple{String, Bool, IntervalsBetween, Bool, typeof(log10), String, Bool}}}}})
    @ Makie ~/.julia/packages/MakieCore/rLlRw/src/recipes.jl:34
 [14] top-level scope
    @ code:9
```

Reply:
I still had to do some tinkering -- I think my local copy of the big arrow table differs a little from yours, but this is something we can fix with a dedicated data package.